akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala

```diff
@@ -0,0 +1,266 @@
+        if (newState.failureCount < maxFailures) {
+          stay using newState
+        } else {
+          goto(Open) using newState
+        }
+
+      case Event(message, state) ⇒
+        log.debug("CLOSED: Sending message {} expecting a response withing timeout {}", message, callTimeout)
+        val currentSender = sender()
+        forwardRequest(message, sender, state)
+        stay
+
+    }
+  }
+
+  when(Open, stateTimeout = resetTimeout.duration) {
+    callSucceededHandling orElse {
+      case Event(StateTimeout, state) ⇒
+        log.debug("Timeout expired for state OPEN, going to half open")
+        goto(HalfOpen) using state.copy(firstHalfOpenMessageSent = false)
+
+      case Event(CallFailed, state) ⇒
+        log.debug("OPEN: Call received a further call failed notification, probably from a previous timed out event, ignoring")
+        stay
+
+      case Event(openNotification @ CircuitOpenFailure(_), _) ⇒
+        log.error("Why did I send message {} to myself?", openNotification)
+        stay
+
+      case Event(message, state) ⇒
+        val failureNotification = failureMap(CircuitOpenFailure(message))
+        log.debug("OPEN: Failing request for message {}, sending failure notification {} to sender {}", message, failureNotification, sender)
+        sender ! failureNotification
+        stay
+
+    }
+  }
+
+  when(HalfOpen) {
+    callSucceededHandling orElse {
+      case Event(TellOnly(message), _) ⇒
+        log.debug("HALF-OPEN: Dropping TellOnly request for message {}", message)
+        stay
+
+      case Event(CallFailed, CircuitBreakerStateData(_, true)) ⇒
+        log.debug("HALF-OPEN: First forwarded call failed returning to OPEN state")
+        goto(Open)
+
+      case Event(CallFailed, CircuitBreakerStateData(_, false)) ⇒
+        log.debug("HALF-OPEN: Call received a further call failed notification, probably from a previous timed out event, ignoring")
+        stay
+
+      case Event(message, state @ CircuitBreakerStateData(_, false)) ⇒
+        log.debug("HALF-OPEN: First message {} received, forwarding it to target {}", message, target)
+        forwardRequest(message, sender, state)
+        stay using state.copy(firstHalfOpenMessageSent = true)
+
+      case Event(message, CircuitBreakerStateData(_, true)) ⇒
+        val failureNotification = failureMap(CircuitOpenFailure(message))
+        log.debug("HALF-OPEN: Failing request for message {}, sending failure notification {} to sender {}", message, failureNotification, sender)
+        sender ! failureNotification
+        stay
+    }
+  }
+
+  def forwardRequest(message: Any, currentSender: ActorRef, state: CircuitBreakerStateData) = {
+    import context.dispatcher
+
+    target.ask(message)(callTimeout).onComplete {
+      case Success(response) ⇒
+        log.debug("Request '{}' has been replied to with response {}, forwarding to original sender {}", message, currentSender)
```

Comment:
Hi @patriknw : I applied all the changes, very good point on the log, I completely overlooked it..